cwl/src/main/scala/wdl4s/cwl/WorkflowStep.scala

```diff
@@ -182,8 +221,29 @@ case class WorkflowStepOutput(id: String)
 
 object WorkflowStep {
 
+  private [cwl] object WorkflowStepInputFold {
+    private [cwl] def emptyRight = WorkflowStepInputFold(Map.empty, Set.empty).asRight[NonEmptyList[String]]
+  }
+  private [cwl] case class WorkflowStepInputFold(stepInputMapping: Map[String, ExpressionNode], generatedNodes: Set[GraphNode]) {
+    def withMapping(stepInput: String, value: ExpressionNode) = this.copy(stepInputMapping = stepInputMapping + (stepInput -> value))
+    def withNewNodes(newNodes: Set[GraphNode]) = this.copy(generatedNodes = generatedNodes ++ newNodes)
+    def withNewNode(newNode: GraphNode) = this.copy(generatedNodes = generatedNodes + newNode)
```

Comment:
ToL since this is a monoid, we could  `|+|` and another instance of `WorkflowStepInputFold`.  This is more terse but I'm unsure about it